packages/dd-trace/src/profiling/profilers/wall.js

```diff
@@ -20,27 +77,82 @@ class NativeWallProfiler {
       process._stopProfilerIdleNotifier = () => {}
     }
 
-    this._record()
+    this._pprof.time.start({
+      intervalMicros: this._samplingIntervalMicros,
+      durationMillis: this._flushIntervalMillis,
+      sourceMapper: this._mapper,
+      customLabels: this._codeHotspotsEnabled,
+      lineNumbers: false })
+
+    if (this._codeHotspotsEnabled) {
+      beforeCh.subscribe(this._enter)
+      enterCh.subscribe(this._enter)
+      afterCh.subscribe(this._exit)
+    }
+
+    this._started = true
+  }
+
+  setLabels (labels) {
+    this._pprof.time.setLabels(labels)
+  }
+
+  _enter () {
+    if (!this._started) return
+
+    const currentSpan = getActiveSpan() || null
+    const currentContext = currentSpan ? currentSpan.context() : null
+
+    if (!currentContext) return
+
+    const startedSpans = getStartedSpans(currentContext)
+    if (!startedSpans || startedSpans.length === 0) return
+    const rootContext = startedSpans[0].context()
+    if (!rootContext) return
+
+    const labels = currentContext ? {
+      'local root span id': rootContext.toSpanId(),
+      'span id': currentContext.toSpanId()
+    } : null
+
+    if (this._endpointCollectionEnabled) {
+      const webServerTags = startedSpans
+        .map(getSpanContextTags)
+        .filter(isWebServerSpan)[0]
+
+      if (webServerTags) {
+        labels['trace endpoint'] = endpointNameFromTags(webServerTags)
+      }
+    }
+
+    this.setLabels(labels)
```

Comment:
I wonder if we could "just" use
```
this.setLabels(currentSpan)
```
and modify the profiler library so we can pass a transformer function to profiler `.start()`. The profiler would then invoke this function on every "labels" object it captured during the serialization step. This way, we'd defer the computation of the actual labels until serialization time, and only do it for spans that were actually captured by the profiler. The correctness of this hinges on whether it will still be possible to correctly retrieve the root span ID and the endpoint tags from the span's context at a later time.